metadata/src/validation.rs

```diff
@@ -0,0 +1,851 @@
+                bytes = hash_hashes(bytes, [hasher.clone() as u8; 32]);
+            }
+            bytes = xor(bytes, get_type_hash(registry, key.id, visited_ids));
+            bytes = xor(bytes, get_type_hash(registry, value.id, visited_ids));
+        }
+    }
+
+    bytes
+}
+
+/// Obtain the hash for a specific storage item, or an error if it's not found.
+pub fn get_storage_hash(
+    metadata: &RuntimeMetadataV15,
+    pallet_name: &str,
+    storage_name: &str,
+) -> Result<[u8; 32], NotFound> {
+    let pallet = metadata
+        .pallets
+        .iter()
+        .find(|p| p.name == pallet_name)
+        .ok_or(NotFound::Pallet)?;
+
+    let storage = pallet.storage.as_ref().ok_or(NotFound::Item)?;
+
+    let entry = storage
+        .entries
+        .iter()
+        .find(|s| s.name == storage_name)
+        .ok_or(NotFound::Item)?;
+
+    let hash = get_storage_entry_hash(&metadata.types, entry, &mut HashSet::new());
+    Ok(hash)
+}
+
+/// Obtain the hash for a specific constant, or an error if it's not found.
+pub fn get_constant_hash(
+    metadata: &RuntimeMetadataV15,
+    pallet_name: &str,
+    constant_name: &str,
+) -> Result<[u8; 32], NotFound> {
+    let pallet = metadata
+        .pallets
+        .iter()
+        .find(|p| p.name == pallet_name)
+        .ok_or(NotFound::Pallet)?;
+
+    let constant = pallet
+        .constants
+        .iter()
+        .find(|c| c.name == constant_name)
+        .ok_or(NotFound::Item)?;
+
+    // We only need to check that the type of the constant asked for matches.
+    let bytes = get_type_hash(&metadata.types, constant.ty.id, &mut HashSet::new());
+    Ok(bytes)
+}
+
+/// Obtain the hash for a specific call, or an error if it's not found.
+pub fn get_call_hash(
+    metadata: &RuntimeMetadataV15,
+    pallet_name: &str,
+    call_name: &str,
+) -> Result<[u8; 32], NotFound> {
+    let pallet = metadata
+        .pallets
+        .iter()
+        .find(|p| p.name == pallet_name)
+        .ok_or(NotFound::Pallet)?;
+
+    let call_id = pallet.calls.as_ref().ok_or(NotFound::Item)?.ty.id;
+
+    let call_ty = metadata.types.resolve(call_id).ok_or(NotFound::Item)?;
+
+    let call_variants = match &call_ty.type_def {
+        TypeDef::Variant(variant) => &variant.variants,
+        _ => return Err(NotFound::Item),
+    };
+
+    let variant = call_variants
+        .iter()
+        .find(|v| v.name == call_name)
+        .ok_or(NotFound::Item)?;
+
+    // hash the specific variant representing the call we are interested in.
+    let hash = get_variant_hash(&metadata.types, variant, &mut HashSet::new());
+    Ok(hash)
+}
+
+/// Obtain the hash representation of a `frame_metadata::v15::PalletMetadata`.
+pub fn get_pallet_hash(
+    registry: &PortableRegistry,
+    pallet: &PalletMetadata<PortableForm>,
+) -> [u8; 32] {
+    // Begin with some arbitrary hash (we don't really care what it is).
+    let mut bytes = hash(&[19]);
```

Comment:
huh, can you elaborate? :)

Reply:
Have added the `MAGIC_PALLET_VALUE` constant and a comment when we use it.
The pallets could basically contain no data, for those cases, we still need to return a hashing value.
And we use this magic arbitrary value to signal that this is an "empty" pallet.